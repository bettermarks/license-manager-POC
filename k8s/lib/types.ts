import { ResourceRequirements } from "../imports/k8s";
/**
 * Stages
 */
export enum Stage {
  LOC = "loc",
  DEV = "dev",
  CI = "ci",
  PRO = "pro",
}

/**
 * Segments
 */
export enum Segment {
  LOC00 = "loc00",
  DEV00 = "dev00",
  CI00 = "ci00",
  CI01 = "ci01",
  PRO00 = "pro00",
}

/**
 * Defines the type defnition of labels in key-value format
 * - Applied to Nodepools when creating K8S cluster
 * - Applied to Deployment definitions at pod spec level when creating a deployment
 */
export type NodeSelector = {
  nodetype: string;
};

export type ApplicationConfig = {
  SEGMENT: string,
  LOG_FORMAT: string;
  LOG_LEVEL: LogLevel;
  APM_URL: string,
  APM_ENABLED: boolean,
  APM_TRANSACTION_SAMPLE_RATE: string,
  GLU_HIERARCHY_PROVIDER_URL: string,
};

/**
 * Deployment configuration
 */
export type DeploymentConfig = {
  /**
   * Migration job resources
   */
  migrationJobResources: ResourceRequirements;
  /**
   * Load fixtures job resources
   */
  loadFixturesJobResources: ResourceRequirements;
  /**
   * API resources
   */
  apiResources: ResourceRequirements;
  /**
   * API replicas
   */
  apiReplicas: number;
};

/**
 * Log level
 */
export enum LogLevel {
  INFO = "INFO",
  WARNING = "WARNING",
  ERROR = "ERROR",
  DEBUG = "DEBUG",
}

/**
 * Namespaces
 */
export enum Namespace {
  DEFAULT = "default",
  LICENSING = "licensing",
}
