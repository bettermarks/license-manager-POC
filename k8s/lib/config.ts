import { Quantity } from "../imports/k8s";

import {
  ApplicationConfig,
  DeploymentConfig,
  LogLevel,
  Segment,
} from "./types";

/**
 * Information to go in secrets
 *
 * DB related:
 * - DATABASE_HOST
 * - DATABASE_PORT
 * - DATABASE_USER
 * - DATABASE_PASSWORD
 * - DATABASE_NAME
 *
 * Application related:
 *
 */

/**
 * Information to go in config map
 */
export const APPLICATION_CONFIG: { [key: string]: ApplicationConfig } = {
  [Segment.LOC00]: {
    SEGMENT: Segment.LOC00,
    LOG_HANDLER: "console",
    LOG_LEVEL: LogLevel.DEBUG,
    DATABASE_USER: "postgres",
    DATABASE_HOST: "localhost",
    DATABASE_PORT: "5432",
    DATABASE_NAME: "licm",
    APM_URL: "https://apm.bettermarks.com",
    APM_ENABLED: false,
    APM_TRANSACTION_SAMPLE_RATE: "0.1",
  },
  [Segment.DEV00]: {
    SEGMENT: Segment.DEV00,
    LOG_HANDLER: "console",
    LOG_LEVEL: LogLevel.DEBUG,
    DATABASE_USER: "postgres",
    DATABASE_HOST: "localhost",
    DATABASE_PORT: "5432",
    DATABASE_NAME: "licm",
    APM_URL: "https://apm.bettermarks.com",
    APM_ENABLED: true,
    APM_TRANSACTION_SAMPLE_RATE: "0.1",
  },
  [Segment.CI00]: {
    SEGMENT: Segment.CI00,
    LOG_HANDLER: "console",
    LOG_LEVEL: LogLevel.DEBUG,
    DATABASE_USER: "postgres",
    DATABASE_HOST: "localhost",
    DATABASE_PORT: "5432",
    DATABASE_NAME: "licm",
    APM_URL: "https://apm.bettermarks.com",
    APM_ENABLED: true,
    APM_TRANSACTION_SAMPLE_RATE: "0.1",
  },
  [Segment.PRO00]: {
    SEGMENT: Segment.PRO00,
    LOG_HANDLER: "console",
    LOG_LEVEL: LogLevel.DEBUG,
    DATABASE_USER: "postgres",
    DATABASE_HOST: "localhost",
    DATABASE_PORT: "5432",
    DATABASE_NAME: "licm",
    APM_URL: "https://apm.bettermarks.com",
    APM_ENABLED: true,
    APM_TRANSACTION_SAMPLE_RATE: "0.1",
  },
};

/**
 * Deployment configuration per segment
 */
export const DEPLOYMENT_CONFIG: {
  [key: string]: DeploymentConfig;
} = {
  [Segment.DEV00]: {
    migrationJobResources: {
      requests: {
        cpu: Quantity.fromNumber(0.1),
        memory: Quantity.fromString("128Mi"),
      },
      limits: {
        cpu: Quantity.fromNumber(1),
        memory: Quantity.fromString("256Mi"),
      },
    },
    loadFixturesJobResources: {
      requests: {
        cpu: Quantity.fromNumber(0.1),
        memory: Quantity.fromString("128Mi"),
      },
      limits: {
        cpu: Quantity.fromNumber(1),
        memory: Quantity.fromString("256Mi"),
      },
    },
    apiResources: {
      requests: {
        cpu: Quantity.fromNumber(0.5),
        memory: Quantity.fromString("512Mi"),
      },
      limits: {
        cpu: Quantity.fromNumber(1),
        memory: Quantity.fromString("1024Mi"),
      },
    },
    apiReplicas: 3,
  },
  [Segment.CI00]: {
    migrationJobResources: {
      requests: {
        cpu: Quantity.fromNumber(0.5),
        memory: Quantity.fromString("128Mi"),
      },
      limits: {
        cpu: Quantity.fromNumber(1),
        memory: Quantity.fromString("256Mi"),
      },
    },
    loadFixturesJobResources: {
      requests: {
        cpu: Quantity.fromNumber(0.5),
        memory: Quantity.fromString("128Mi"),
      },
      limits: {
        cpu: Quantity.fromNumber(1),
        memory: Quantity.fromString("256Mi"),
      },
    },
    apiResources: {
      requests: {
        cpu: Quantity.fromNumber(1),
        memory: Quantity.fromString("512Mi"),
      },
      limits: {
        cpu: Quantity.fromNumber(2),
        memory: Quantity.fromString("1024Mi"),
      },
    },
    apiReplicas: 3,
  },
  [Segment.PRO00]: {
    migrationJobResources: {
      requests: {
        cpu: Quantity.fromNumber(0.5),
        memory: Quantity.fromString("128Mi"),
      },
      limits: {
        cpu: Quantity.fromNumber(1),
        memory: Quantity.fromString("256Mi"),
      },
    },
    loadFixturesJobResources: {
      requests: {
        cpu: Quantity.fromNumber(0.5),
        memory: Quantity.fromString("128Mi"),
      },
      limits: {
        cpu: Quantity.fromNumber(1),
        memory: Quantity.fromString("256Mi"),
      },
    },
    apiResources: {
      requests: {
        cpu: Quantity.fromNumber(1),
        memory: Quantity.fromString("512Mi"),
      },
      limits: {
        cpu: Quantity.fromNumber(2),
        memory: Quantity.fromString("1024Mi"),
      },
    },
    apiReplicas: 3,
  },
};
