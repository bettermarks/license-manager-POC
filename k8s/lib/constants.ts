import { NodeSelector } from "./types";

export const APP_NODE_POOL_LABELS: NodeSelector = {
  nodetype: "application",
};

/**
 * Name of the secret which is used by service chart to look up application secret.
 * This secret is created by AWS Secrets helper in bm-operations project. https://github.com/bettermarks/bm-operations/blob/master/cdk8s/apps/licensing.ts
 */
export const APPLICATION_SECRET = "licensing-secret";
/**
 * Name of the secret which is used by service chart to look up Postgres DB credentials.
 * This secret is created  by AWS Secrets helper in bm-operations project. https://github.com/bettermarks/bm-operations/blob/master/cdk8s/apps/licensing.ts
 */
export const POSTGRES_SECRET = "licensing-postgres-secret";
/**
 * Name of the secret used by Service Account to pull image from AWS ECR registry.
 * This secret is created by AWS Registry Helper Cron. https://github.com/bettermarks/bm-operations/blob/master/cdk8s/apps/licensing.ts
 */
export const REGISTRY_CREDENTIALS = "registry-credentials";
/**
 * Name of the image used for postgres database.
 */
export const POSTGRES_IMAGE = "postgres:14";
