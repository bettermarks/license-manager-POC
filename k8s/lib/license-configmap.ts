import { ConfigMap } from "cdk8s-plus-24";
import { Construct } from "constructs";
import { ApplicationConfig } from "./types";

export type LicenseConfigMapProps = {
  name: string;
  appConfig: ApplicationConfig;
  namespace?: string;
};

export class LicenseConfigMap extends Construct {
  readonly configMap: ConfigMap;

  constructor(scope: Construct, id: string, props: LicenseConfigMapProps) {
    super(scope, id);
    const { name, namespace, appConfig } = props;

    this.configMap = new ConfigMap(this, `${name}-configmap`, {
      metadata: {
        name: `${name}-configmap`,
        namespace,
      },
    });
    for (const [key, value] of Object.entries(appConfig)) {
      if (Array.isArray(value) || typeof value === "boolean") {
        this.configMap.addData(key, JSON.stringify(value));
      } else {
        this.configMap.addData(key, value);
      }
    }
  }
}