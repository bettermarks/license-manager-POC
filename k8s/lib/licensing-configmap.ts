import { ConfigMap } from "cdk8s-plus-24";
import { Construct } from "constructs";
import { ApplicationConfig } from "./types";

export type LicensingConfigMapProps = {
  name: string;
  appConfig: ApplicationConfig;
  namespace?: string;
};

export class LicensingConfigMap extends Construct {
  readonly configMap: ConfigMap;

  constructor(scope: Construct, id: string, props: LicensingConfigMapProps) {
    super(scope, id);
    const { name, namespace, appConfig } = props;

    this.configMap = new ConfigMap(this, `${name}-configmap`, {
      metadata: {
        name: `${name}-configmap`,
        namespace,
      },
    });
    for (const [key, value] of Object.entries(appConfig)) {
      if (Array.isArray(value) || typeof value === "boolean") {
        this.configMap.addData(key, JSON.stringify(value));
      } else {
        this.configMap.addData(key, value);
      }
    }
  }
}
