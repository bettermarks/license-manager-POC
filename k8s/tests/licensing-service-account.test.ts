import { Testing } from "cdk8s";
import { LicensingServiceAccount } from "../lib/service-account";

describe("Licensing-service-account", () => {
    test("should match with snapshot", () => {
        const app = Testing.app();
        const chart = new LicensingServiceAccount(app, "test-chart", {
            imagePullSecrets: ["ImagePullSecret"],
            namespace: "dummy-namespace",
            name: "service-account",
        });
        const results = Testing.synth(chart);
        expect(results).toMatchSnapshot();
    });
});
