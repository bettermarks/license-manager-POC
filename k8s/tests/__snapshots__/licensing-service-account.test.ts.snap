// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Licensing-service-account should match with snapshot 1`] = `
Array [
  Object {
    "apiVersion": "v1",
    "imagePullSecrets": Array [
      Object {
        "name": "ImagePullSecret",
      },
    ],
    "kind": "ServiceAccount",
    "metadata": Object {
      "name": "service-account",
      "namespace": "dummy-namespace",
    },
  },
  Object {
    "apiVersion": "rbac.authorization.k8s.io/v1",
    "kind": "Role",
    "metadata": Object {
      "name": "service-account",
      "namespace": "dummy-namespace",
    },
    "rules": Array [
      Object {
        "apiGroups": Array [
          "",
        ],
        "resources": Array [
          "pods",
        ],
        "verbs": Array [
          "get",
          "list",
          "watch",
        ],
      },
    ],
  },
  Object {
    "apiVersion": "rbac.authorization.k8s.io/v1",
    "kind": "RoleBinding",
    "metadata": Object {
      "name": "service-account",
      "namespace": "dummy-namespace",
    },
    "roleRef": Object {
      "apiGroup": "rbac.authorization.k8s.io",
      "kind": "Role",
      "name": "service-account",
    },
    "subjects": Array [
      Object {
        "kind": "ServiceAccount",
        "name": "service-account",
      },
    ],
  },
]
`;
